app/validation/validator.py

```diff
@@ -1734,185 +1723,149 @@ def _get_dicts_with_key(self, input_data, key_name):
-                    current_block_id,
-                    placeholder_definition,
-                    answers_with_parent_ids,
-                    answer_ids_to_validate,
-                )
-            )
-            errors.extend(
-                self._validate_placeholder_metadata_ids(
-                    valid_metadata_ids,
-                    metadata_ids_to_validate,
-                    placeholder_definition["placeholder"],
-                )
-            )
-
             if transforms:
                 errors.extend(
                     self._validate_placeholder_transforms(transforms, current_block_id)
                 )
 
-        placeholder_differences = set(placeholders_in_string) - set(
-            placeholder_definition_names
-        )
+        placeholder_differences = placeholders_in_string - placeholder_definition_names
         if placeholder_differences:
+            try:
+                text = placeholder_object["text"]
+            except KeyError:
+                text = placeholder_object["text_plural"]["forms"]["other"]
             errors.append(
                 self._error_message(
                     "Placeholders in '{}' don't match definitions. Missing '{}'".format(
-                        placeholder_object["text"], placeholder_differences
+                        text, placeholder_differences
                     )
                 )
             )
 
         return errors
 
-    def _validate_placeholders(
-        self, block_json, answers_with_parent_ids, valid_metadata_ids, current_block_id
-    ):
+    def _validate_placeholders(self, block_json):
         errors = []
         strings_with_placeholders = self._get_dicts_with_key(block_json, "placeholders")
         for placeholder_object in strings_with_placeholders or []:
             errors.extend(
-                self._validate_placeholder_object(
-                    answers_with_parent_ids,
-                    valid_metadata_ids,
-                    placeholder_object,
-                    current_block_id,
-                )
+                self._validate_placeholder_object(placeholder_object, block_json["id"])
             )
 
         return errors
 
-    def _validate_list_exists(self, list_name):
-        if list_name not in self._list_names:
-            msg = f"for_list '{list_name}' is not populated by any ListCollector blocks"
-            return [self._error_message(msg)]
-        return []
-
-    def _validate_relationship_collector_answers(self, answers):
-        one_answer_msg = "RelationshipCollector contains more than one answer."
-        answer_type_msg = "Only answers of type Relationship are valid in RelationshipCollector blocks."
-
+    def _validate_source_references(
+        self, block_json, answers_with_parent_ids, valid_metadata_ids
+    ):
         errors = []
+        source_references = self._get_dicts_with_key(block_json, "identifier")
+        for source_reference in source_references or []:
```

Comment:
if self._get_dict_with_key returned an empty iterable by default then the `or []` pattern which is used frequently wouldn't be needed